### PR TITLE
Fail the test when there are active handles in the loop at the end

### PR DESCRIPTION
**Pull request by foolswood**
Just to make sure there are no outstanding timers (or half finished other things).

I found it helpful to keep tests self contained.

Will tidy and test further if you're interested. Just wondered what you make of such a thing.

### CODE REVIEW

**Comment by Martiusweb** [comment]
Hi,

Thanks for your contribution. I would like to add several sanity checks at the end of a test, checking that there are no remaining scheduled callbacks is one of them (an other example is #3 - for a selector event loop). However, I'm not sure these checks should cause a test to fail by default (at least it must be configurable).

What do you think of replacing something like ``@asynctest.ignore_loop`` by something like:

    @asynctest.fail_on(
        unused_loop=False,
        active_handles=True,
        # anything else we see fit
    )

where the decorator allows us to add as many optional checks we see fit.

Currently, you can do this assertion without changing asynctest in the ``TestCase.tearDown()``.

**Comment by foolswood** [comment]
Apologies for the delay, I completely forgot about this, if it happens again please nag me to get on it.

**Comment by foolswood** [comment]
Needs more tests and such, but is this a better shape?

File: asynctest/case.py [hunk 1]
```diff
@@ -306,19 +316,27 @@ def debug(self):
...
-                getattr(method, "__asynctest_ignore_loop__", False)):
-            return
-
-        if not self.loop.__asynctest_ran:
-            self.fail("Loop did not run during the test")
+        else:
+            if fail_checks['unused_loop']:
+                if not self.loop.__asynctest_ran:
+                    self.fail("Loop did not run during the test")
+            if fail_checks['active_handles']:
```
Comment (Martiusweb) [inline asynctest/case.py]: the "unused loop" test is useless when the test method is a coroutine (the loop will run the coroutine), but "active handles" can (and should) be tested in both cases (coroutine or function).

File: asynctest/case.py [hunk 2]
```diff
@@ -172,13 +172,23 @@ def _init_loop(self):
 
         self.loop = self._patch_loop(self.loop)
 
+    @staticmethod
+    def _scheduled_handle_alive(handle):
+        cancelled_timer_handle = (
+            isinstance(scheduled_thing, asyncio.TimerHandle) and
```
Comment (Martiusweb) [inline asynctest/case.py]: Is `scheduled_thing` actually `handle`?
I prefer the name `scheduled_timer_handle_is_alive()`: the functions contains a verb and is more precise (as it returns `False` for `asyncio.Handle`).

File: asynctest/case.py [hunk 3]
```diff
@@ -172,13 +172,23 @@ def _init_loop(self):
...
+            scheduled_thing._cancelled)
+        return not cancelled_timer_handle
+
+    @property
+    def _live_handles(self):
```
Comment (Martiusweb) [inline asynctest/case.py]: `_live_timer_handles()` (as above)

File: asynctest/case.py [hunk 4]
```diff
@@ -346,3 +364,27 @@ def ignore_loop(test):
     """
     test.__asynctest_ignore_loop__ = True
```
Comment (Martiusweb) [inline asynctest/case.py]: It can be removed.

**Comment by Martiusweb** [comment]
Hi, I added a few comments, but that's the idea, thanks!
One important remark is that the check should be done if the test method is a coroutine (currently it's performed only if it's a function).

I'll take a more thorough look when the feature works and you have tests. The code in `tests/test_case.py` is a bit messy because of the mixins. If you're lost, you can write tests in a separate class and I'll refactor them later.

**Comment by foolswood** [comment]
This seems to be going stale, anything to do on it?

**Comment by Martiusweb** [comment]
Hi,

I implemented the @fail_on mechanism for asynctest 0.8.0, there is an example of how to add a check here:
https://github.com/Martiusweb/asynctest/blob/master/asynctest/selector.py#L340

It should not be too hard to adapt your patch. Do you want to do it?

**Comment by foolswood** [comment]
How about that?

File: test/test_case.py [hunk 5]
```diff
@@ -230,6 +230,18 @@ def runTest(self):
                         getattr(case, method)()
                         self.assertTrue(case.ran)
 
+    def test_fails_when_future_has_scheduled_calls(self):
+        class CruftyTest(asynctest.TestCase):
+            @asynctest.fail_on(active_handles=True, unused_loop=False)
+            def runTest(instance):
+                instance.loop.call_later(5, lambda: None)
+
+        with self.subTest(test=CruftyTest):
```
Comment (Martiusweb) [inline test/test_case.py]: actually, this subTest is subTest(method="debug")

File: test/test_case.py [hunk 6]
```diff
@@ -230,6 +230,18 @@ def runTest(self):
...
+            with self.assertRaisesRegex(AssertionError, 'unfinished work'):
+                CruftyTest().debug()
+            result = CruftyTest().run()
```
Comment (Martiusweb) [inline test/test_case.py]: ... and this one subTest(method="run")

**Comment by Martiusweb** [comment]
Just a small comment on the test and it seems good to merge.

File: test/test_case.py (hunk 6 above)
Comment (Martiusweb) [inline test/test_case.py]: ... and this one subTest(method="run")

**Comment by Martiusweb** [comment]
Thanks, I'll do a new release sooon.

**Pull request merged**